server/src/main/java/org/elasticsearch/search/aggregations/metrics/ScriptedMetricAggregator.java

```diff
@@ -90,6 +91,7 @@ public InternalAggregation buildAggregation(long owningBucketOrdinal) {
         } else {
             aggregation = aggState;
         }
+        StreamOutput.checkWriteable(aggregation);
```

Comment:
I wonder if we could/should combine this with `org.elasticsearch.common.util.CollectionUtils#ensureNoSelfReferences(java.lang.Object, java.lang.String)` somehow? It seems like technically we would want to use this in all the spots that we currently also use that method to validate script returns (maybe I'm missing some detail and we wouldn't need to do this kind of validation in other script return spots and only need the no-self-references there for some reason?). That way we'd save iterating through the object to validate twice?

Reply:
I'm not sure we need both in all cases. ensureNoSelfReferences guarantees we can do any operation on the object in question (not resulting in an infinite loop). Checking if the object is writeable is only needed when we plan to write the object to transfer to another node. For example, with scripted metric aggs, we don't every write the node local results from init or map scripts, only the output of combine here that would be transferred to the coordinating node for the final reduce.

Reply:
Ok makes sense then :) => I think the approach is fine.